File: sdk/desktop/macos/Tests/KsealDesktopTests/ProvidersTests.swift
import XCTest
@testable import KsealDesktop

final class ProvidersTests: XCTestCase {

    private func makeDir() -> URL {
        let dir = FileManager.default.temporaryDirectory
            .appendingPathComponent("kseal-test-\(UUID().uuidString)", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    func testStorageScopeIsStableAndTenantAppSpecific() {
        let a = StorageScope.component(tenantId: "tenant-1", appId: "app-1")
        XCTAssertEqual(a, StorageScope.component(tenantId: "tenant-1", appId: "app-1"))
        XCTAssertNotEqual(a, StorageScope.component(tenantId: "tenant-2", appId: "app-1"))
        XCTAssertNotEqual(a, StorageScope.component(tenantId: "tenant-1", appId: "app-2"))
        XCTAssertNotNil(a.range(of: "^[0-9a-f]+$", options: .regularExpression))
    }

    func testProofKeyIsStableAcrossInstances() {
        let dir = makeDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let key1 = DefaultProofKeyProvider(directory: dir).proofKey()
        let key2 = DefaultProofKeyProvider(directory: dir).proofKey()
        XCTAssertEqual(key1.count, 32)
        XCTAssertEqual(key1, key2)
    }

    func testCreateOrReadExistingKeepsTheFirstWriter() {
        let dir = makeDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let url = dir.appendingPathComponent("proof.key")
        let first = DefaultProofKeyProvider.createOrReadExisting(at: url, candidate: Data([1, 1, 1]))
        let second = DefaultProofKeyProvider.createOrReadExisting(at: url, candidate: Data([2, 2, 2]))
        XCTAssertEqual(first, Data([1, 1, 1]))
        XCTAssertEqual(second, first) // loser adopts the winner's bytes
    }

    func testInstallIdentityHashIsStable() {
        let dir = makeDir()
        defer { try? FileManager.default.removeItem(at: dir) }
        let h1 = InstallIdentity(directory: dir).tenantScopedHash(tenantId: "t", appId: "a")
        let h2 = InstallIdentity(directory: dir).tenantScopedHash(tenantId: "t", appId: "a")
        XCTAssertEqual(h1, h2)
        XCTAssertFalse(h1.isEmpty)
    }
}

File: sdk/desktop/windows/tests/Kseal.Desktop.Tests/ProvidersTests.cs
using Xunit;

namespace Kseal.Desktop.Tests;

public class ProvidersTests : IDisposable
{
    private readonly string _dir = Path.Combine(Path.GetTempPath(), $"kseal-test-{Guid.NewGuid():N}");

    public ProvidersTests() => Directory.CreateDirectory(_dir);

    public void Dispose()
    {
        try { Directory.Delete(_dir, recursive: true); } catch (IOException) { }
        GC.SuppressFinalize(this);
    }

    [Fact]
    public void StorageScopeIsStableAndTenantAppSpecific()
    {
        string a = StorageScope.Component("tenant-1", "app-1");
        Assert.Equal(a, StorageScope.Component("tenant-1", "app-1"));       // deterministic
        Assert.NotEqual(a, StorageScope.Component("tenant-2", "app-1"));    // tenant matters
        Assert.NotEqual(a, StorageScope.Component("tenant-1", "app-2"));    // app matters
        // No path separators / unsafe chars (lowercase hex only).
        Assert.Matches("^[0-9a-f]+$", a);
    }

    [Fact]
    public void AtomicWriteReplacesExistingContent()
    {
        string path = Path.Combine(_dir, "config.bin");
        AtomicFile.Write(path, [1, 2, 3]);
        AtomicFile.Write(path, [4, 5]);
        Assert.Equal([4, 5], File.ReadAllBytes(path));
        // No temp files left behind.
        Assert.Empty(Directory.GetFiles(_dir, "*.tmp"));
    }

    [Fact]
    public void CreateOrReadExistingKeepsTheFirstWriter()
    {
        string path = Path.Combine(_dir, "proof.key");
        byte[] first = AtomicFile.CreateOrReadExisting(path, [1, 1, 1]);
        byte[] second = AtomicFile.CreateOrReadExisting(path, [2, 2, 2]);
        Assert.Equal([1, 1, 1], first);
        Assert.Equal(first, second); // loser adopts the winner's bytes
    }

    [Fact]
    public void ProofKeyAndInstallIdAreStableAcrossInstances()
    {
        byte[] key1 = new DefaultProofKeyProvider(_dir).ProofKey();
        byte[] key2 = new DefaultProofKeyProvider(_dir).ProofKey();
        Assert.Equal(32, key1.Length);
        Assert.Equal(key1, key2);

        string hash1 = new InstallIdentity(_dir).TenantScopedHash("t", "a");
        string hash2 = new InstallIdentity(_dir).TenantScopedHash("t", "a");
        Assert.Equal(hash1, hash2);
    }
}
